src/platform/cocoamain.mm

```diff
@@ -1218,6 +1218,227 @@ - (IBAction)preferences:(id)sender {
     [NSApp stop:nil];
 }
 
+#ifdef HAVE_SPACEWARE
+/*
+ * Normally we would just link to the 3DconnexionClient framework.
+ * We don't want to (are not allowed to) distribute the official
+ * framework, so we're trying to use the one installed on the users
+ * computer. There are some different versions of the framework,
+ * the official one and re-implementations using an open source driver
+ * for older devices (spacenav-plus). So weak-linking isn't an option,
+ * either. The only remaining way is using CFBundle to dynamically
+ * load the library at runtime, and also detect its availability.
+ *
+ * We're also defining everything needed from the 3DconnexionClientAPI,
+ * so we're not depending on the API headers.
+ */
+
+#pragma pack(push,2)
+
+enum {
+    kConnexionClientModeTakeOver = 1,
+    kConnexionClientModePlugin = 2
+};
+
+#define kConnexionMsgDeviceState '3dSR'
+#define kConnexionMaskButtons 0x00FF
+#define kConnexionMaskAxis 0x3F00
+
+typedef struct {
+    uint16_t version;
+    uint16_t client;
+    uint16_t command;
+    int16_t param;
+    int32_t value;
+    UInt64 time;
+    uint8_t report[8];
+    uint16_t buttons8;
+    int16_t axis[6];
+    uint16_t address;
+    uint32_t buttons;
+} ConnexionDeviceState, *ConnexionDeviceStatePtr;
+
+#pragma pack(pop)
+
+typedef void (*ConnexionAddedHandlerProc)(io_connect_t);
+typedef void (*ConnexionRemovedHandlerProc)(io_connect_t);
+typedef void (*ConnexionMessageHandlerProc)(io_connect_t, natural_t, void *);
+
+typedef OSErr (*InstallConnexionHandlersProc)(ConnexionMessageHandlerProc, ConnexionAddedHandlerProc, ConnexionRemovedHandlerProc);
+typedef void (*CleanupConnexionHandlersProc)(void);
+typedef UInt16 (*RegisterConnexionClientProc)(UInt32, UInt8 *, UInt16, UInt32);
+typedef void (*UnregisterConnexionClientProc)(UInt16);
+
+@interface FrameworkWrapper : NSObject {
+    InstallConnexionHandlersProc installConnexionHandlers;
+    CleanupConnexionHandlersProc cleanupConnexionHandlers;
+    RegisterConnexionClientProc registerConnexionClient;
+    UnregisterConnexionClientProc unregisterConnexionClient;
+
+    CFBundleRef cfBundle;
+}
+
+- (OSErr) installConnexionHandlers:(ConnexionMessageHandlerProc)message
+                             Added:(ConnexionAddedHandlerProc)added
+                           Removed:(ConnexionRemovedHandlerProc)removed;
+
+- (void) cleanupConnexionHandlers;
+
+- (UInt16) registerConnexionClient:(UInt32)signature Name:(UInt8 *)name
+                              Mode:(UInt16)mode Mask:(UInt32)mask;
+
+- (void) unregisterConnexionClient:(UInt16)client;
+@end
+
+@implementation FrameworkWrapper
+- (id)init {
+    self = [super init];
+    NSString *bundlePath = @"/Library/Frameworks/3DconnexionClient.framework";
```

Comment:
That is a good question, I've spent some time searching but didn't find anything. I suppose it could be in other locations, like `/System/Library/Frameworks`, but as far as I am aware the official 3DconnexionClient is always installed in `/Library/Frameworks`.

NSBundle can be used to get the paths for frameworks inside of the application bundle, but I can't find anything for the system level.